### PR TITLE
Remove YSort node, move its functionality into Node2D

### LINKED ISSUE #1915: Document new nested YSort feature
@bojidar-bg implemented a new, long awaited feature in YSort with godotengine/godot#23407.

The docs (both classref and, if relevant, tutorials) and demos should be updated to reflect that and show how to use it.

### LINKED ISSUE #2286: Ysort practical example needed
We need [a practical example like this](https://www.reddit.com/r/godot/comments/b4b22x/ysorting/ej5qbam?utm_source=share&utm_medium=web2x) for the Ysort node in the [docs](https://docs.godotengine.org/en/3.1/classes/class_ysort.html). Without it, I would also assume Nodes would have [the same behaviour OP assumes in his post](https://www.reddit.com/r/godot/comments/b4b22x/ysorting/). In fact I have struggled with this before and when I asked how to Ysort, people have recommended having all sorted nodes as children as of a tilemap. (which does not allow grouping either, so I did not understand the purpose of Ysort very well) So to avoid confusion and wrong advice, let's add this great example:

```
Node (main)
-> YSort (World)
   -> Player
   -> YSort (Environment)
      -> Rock
      -> Tree
   -> YSort (Enemies)
      -> Enemy1
      -> Enemy2
```

### PR DESCRIPTION
**Pull request by andriyDev**
Currently, all a YSort node does is control calls to RenderingServer.canvas_item_set_sort_children_by_y. By deleting YSort and moving these calls to Node2D, we can now have any node be treated as a YSort (if it wants). The main advantage of this is that we can now have "deep" y-sorting if desired. In order to y-sort children at the moment, all intermediate nodes must also be YSort nodes, which is very limiting. By making all Node2D's effectively YSort nodes, we can have deep y-sorting anywhere.

In addition, we replace the `cell_y_sort` property in TileMap with `y_sort_enabled`.

Discussion of this proposal is offered at [godotengine/godot-proposals/issues/1237](https://github.com/godotengine/godot-proposals/issues/1237), similar to Proposition A mentioned in the original proposal.

_Edit by godotengine/documentation:_
Closes godotengine/godot-docs#2286
Closes godotengine/godot-docs#1915

### CODE REVIEW

File: doc/classes/Node2D.xml
Comment (NathanLovato) [inline doc/classes/Node2D.xml:160]:
```suggestion
			If [code]true[/code], child nodes with the lowest Y position are drawn before those with a higher Y position. If [code]false[/code], Y-sorting is disabled. Y-sorting only affects children that inherit from [CanvasItem].
			You can nest nodes with Y-sorting. Children Y-sorted nodes are sorted in the same space as the parent Y-sort. This feature allows you to organize a scene better or divide it into multiple ones without changing your scene tree.
```

**Review by NathanLovato — CHANGES_REQUESTED** [review]
This feature sounds great. The docs could be clearer and follow our writing guidelines better. I've left a suggestion. 

Note for later: If this gets merged, we should write a manual talking about Z-order that we can link from Node2D's class reference.

File: doc/classes/Node2D.xml
Comment (andriyDev) [reply doc/classes/Node2D.xml:160]: This is a much better description, thank you! I was unsure what the best way to document the feature was, it is mostly just a copy of the original YSort docs, with slight modifications. I definitely agree that a manual describing the feature would be more clear than a member description. I will update the PR as soon as I can.

Comment (andriyDev) [reply doc/classes/Node2D.xml]: I have updated the PR.

**Review by NathanLovato — APPROVED** [review]
Approving the documentation

**Comment by KoBeWi** [comment]
We discussed this PR briefly on IRC some time ago and the change is good, but the YSort node should be deprecated instead of removed (to not break project migrating to 4.0). For now it would be the same as Node2D, but with `y_sort_enabled` being true by default, and it should have a deprecation warning saying that it will be removed (like AnimationTreePlayer had). Then the YSort will be removed in 4.1.

**Comment by andriyDev** [comment]
> We discussed this PR briefly on IRC some time ago and the change is good, but the YSort node should be deprecated instead of removed (to not break project migrating to 4.0). For now it would be the same as Node2D, but with `y_sort_enabled` being true by default, and it should have a deprecation warning saying that it will be removed (like AnimationTreePlayer had). Then the YSort will be removed in 4.1.

This is a good point, I will make the necessary changes when I can.

**Comment by andriyDev** [comment]
I have updated the PR to have YSort deprecated!

File: doc/classes/Node2D.xml
Comment (akien-mga) [inline doc/classes/Node2D.xml:162]: Those are normally ordered alphabetically, so it could come before `z_as_relative`. Did you add it here manually or did `doctool` place it there? (in which case that would be a bug)

File: doc/classes/Node2D.xml
Comment (akien-mga) [inline doc/classes/Node2D.xml:161]: Refers both to "child nodes" and "children [Y-sorted] nodes", not sure which one is most correct in English (child vs children as adjective).

File: doc/classes/YSort.xml
Comment (akien-mga) [inline doc/classes/YSort.xml:4]:
```suggestion
		[i]Deprecated.[/i] Sort all child nodes based on their Y positions. This has been replaced by the [member Node2D.y_sort_enabled] property.
```

File: doc/classes/YSort.xml
Comment (akien-mga) [inline doc/classes/YSort.xml:9]:
```suggestion
		This functionality has been replaced by the [member Node2D.y_sort_enabled] property.
```

File: doc/classes/YSort.xml
Comment (akien-mga) [inline doc/classes/YSort.xml:17]: This should be preserved if we're not outright remove the node, so existing scripts still work (and thus it should change `Node2D.y_sort_enabled` when toggled).

File: scene/2d/tile_map.cpp
Comment (akien-mga) [inline scene/2d/tile_map.cpp:1831]: Note: If we're choosing to preserve compatibility, this should likely be kept as a deprecated property too.. But I'm starting to question whether asking to keep YSort as deprecated is a good idea. This is for 4.0 and we've broken the API is way more significant ways, not sure that it's worth it to keep stubs where it should be fairly easy for users (if documented) to review their scenes for uses a "YSort" named nodes that changed to `Node2D` type, and tick a property.

**Comment by andriyDev** [comment]
I pushed a change which actually used doctool to generate the docs, I didn't realize that that was a tool! The docs are now updated correctly, including the YSort doc which contains the new defaulted value.

File: scene/2d/tile_map.cpp
Comment (NathanLovato) [reply scene/2d/tile_map.cpp:1831]: Considering 4.0 is the time to break the API, that probably Godot users will have to do some work to port their projects, even with some tools to help them (new GDScript, many changes in resources, and more to come), that'd be a good time to remove YSort, considering the feature completely supersedes it.

As a game dev, I'd personally prefer an article in the docs detailing replaced features and steps to update your project than a deprecation warning. Also, going from Godot 3 to 4 tells me my project aren't compatible with the new version anyway so I wouldn't even try to port e.g. a commercial project.

File: doc/classes/Node2D.xml
Comment (andriyDev) [reply doc/classes/Node2D.xml]: I have fixed this by actually using doctool, didn't even know that existed!

File: doc/classes/YSort.xml
Comment (andriyDev) [reply doc/classes/YSort.xml]: Good point, I didn't realize that was the notation for that. Fixed

File: scene/2d/tile_map.cpp
Comment (andriyDev) [reply scene/2d/tile_map.cpp]: Reduz said in IRC that we should still deprecate YSort since we still want to be able to open Godot 3 scenes in Godot 4. As such I've properly implemented the deprecated YSort this time. Let me know if there are any changes!

File: doc/classes/YSort.xml
Comment (andriyDev) [reply doc/classes/YSort.xml]: Fixed, I also kept both functions set_sort_enabled and is_sort_enabled in order to ensure full compatibility while it is deprecated. They just immediately call set_y_sort_enabled and is_y_sort_enabled respectively.

**Comment by andriyDev** [comment]
I've made the deprecated YSort fully backwards compatible (how it should have been, my mistake) and fixed some doc formatting issues.

File: doc/classes/TileMap.xml
Comment (asheraryam) [inline doc/classes/TileMap.xml:283]: ~~This future is~~ In the future this is

**Comment by avencherus** [comment]
Since the node would remain as deprecated, was wondering what might be the consensus of a 3.2 port of this.

In certain projects I use YSort quite a bit, and often mixed in with some complicated layouts.

This would add a lot of flexibility to how the scenes could be laid out, and get rid of a lot of nested YSorts.  Also would make certain things possible when direct nesting can't be achieved.  Alleviating the need for awkward work arounds.

Introducing it sooner than later may also start the process of migrating away from it.

File: doc/classes/Node2D.xml
Comment (andriyDev) [reply doc/classes/Node2D.xml]: This is a good point, I decided that "Child Y-sorted nodes" sounds a little better. and also more consistent.

File: doc/classes/YSort.xml
Comment (andriyDev) [reply doc/classes/YSort.xml]: I have updated this as well, I did not know the notation before.

File: doc/classes/TileMap.xml
Comment (andriyDev) [reply doc/classes/TileMap.xml:283]: UGH Why am I typo-city recently. Thanks, good catch!

**Comment by andriyDev** [comment]
> Since the node would remain as deprecated, was wondering what might be the consensus of a 3.2 port of this.
> 
> In certain projects I use YSort quite a bit, and often mixed in with some complicated layouts.
> 
> This would add a lot of flexibility to how the scenes could be laid out, and get rid of a lot of nested YSorts. Also would make certain things possible when direct nesting can't be achieved. Alleviating the need for awkward work arounds.
> 
> Introducing it sooner than later may also start the process of migrating away from it.

Based on my discussions with Akien and reduz in IRC, the consensus seems to be that the better solution would simply be to remove YSort and then just make a compatibility alias so 3.* scenes will have YSorts replaced with Node2Ds in 4.0 (and they will need to manually be set to y_sort_enabled). I would recommend to just go through and add this change yourself to a personal build if you have one, it's a pretty small change all things considered.

**Comment by andriyDev** [comment]
Fixed a typo in the TileMap docs.

**Comment by avencherus** [comment]
> I would recommend to just go through and add this change yourself to a personal build if you have one, it's a pretty small change all things considered.

Thanks, yeah something I'll definitely be including.  Appreciate this commit.  👍

**Comment by avencherus** [comment]
@andriyDev Tested this with the master, found a problem.  Nested Sprites that use the property don't render their textures at depth 1 and lower.  May also affect other texture/canvas related nodes.

image

https://github.com/godotengine/godot/blob/master/servers/rendering/rendering_server_canvas.cpp ysort logic might need to be updated.  Probably skipping the parent.

Here's an MRP: [YSort Issue.zip](https://github.com/godotengine/godot/files/5345144/YSort.Issue.zip)

**Comment by andriyDev** [comment]
Thanks for notifying me about this issue! I'm gonna see what I can do about it (when I have the time). Until then, this PR probably shouldn't be merged. While this still is a useful feature (for example y-sorting children under a Navigation2D node), it does create unclear behaviour which we should try to avoid.

**Comment by avencherus** [comment]
@andriyDev Looking at recent work, @bojidar-bg might know more about these interactions.

**Comment by andriyDev** [comment]
@avencherus I've already found some interesting parts in the _cull_canvas_item function. It looks like if a canvas item is ysort and its child is ysort it terminates. If you comment that out the behaviour is very strange (very hard to explain). It looks like I may need to just parse through all of it to understand why things are the way they are. If I need help I'll definitely ask @bojidar-bg  :)

**Comment by bojidar-bg** [comment]
Looking at the code, the likely culprit is the `(ci->sort_y && child_items[i]->sort_y)` check: https://github.com/godotengine/godot/blob/fe6109101f4b8f7383a2dd03546bcbda9cbd042b/servers/rendering/rendering_server_canvas.cpp#L171-L172 https://github.com/godotengine/godot/blob/fe6109101f4b8f7383a2dd03546bcbda9cbd042b/servers/rendering/rendering_server_canvas.cpp#L214-L215

The main reason that check exists is to prevent `_cull_canvas_item` to recurse, collecting the same YSorted nodes multiple times. I suppose it can be removed if the code related to drawing a canvas item (without its children) is separated from the code related to iterating through canvas items, perhaps by splitting the function in two or by introducing a flag.

**Comment by avencherus** [comment]
Was also glancing it from 3.2 as `_render_canvas_item()`.  I see, they won't get added to the z_list.

**Comment by andriyDev** [comment]
@bojidar-bg @avencherus @akien-mga I am unsure how I should make the behaviour... The currently pushed revision works overall, however it has a large caveat - the root of the Y-sorting is not itself Y-sorted. Ex: if you have 3 sprites in a single chain, the bottom 2 sprites will be Y-sorted, but the root will always render behind (except if Show Behind Parent is true). This behaviour is very easy based on the current implementation.

This is not very intuitive behaviour to me. Ideally, a parent would also be Y-sorted amongst its children, so the root would be able to render ahead or behind its children based on the children's vertical height. I believe I can make this modification, although it would be more code. In addition, it does make the "Show Behind Parent" property on children a little unclear, and would almost certainly need to be ignored during Y-sort amongst the parent, which I think is fine. Please let me know of any opinions you all have!

**Comment by andriyDev** [comment]
Marking this commit just above as the "simple" implementation - not Y-sorting parent amongst its children.

**Comment by andriyDev** [comment]
This is the commit for the "complex" implementation - Y-sorting parent amongst its children. This required a reorganization of how canvas items are culled and organized. Specifically, splitting off a function to "attach" canvas items to the list for drawing, and using that to reorder how these "attach" calls are issued.

If the opinion is to use the other method, I have a branch to easily push it onto the PR.

**Comment by avencherus** [comment]
You're right, very good point.

It's a bit vague how `behind_parent` would or should work.  Looking at the current behavior with YSort nodes, it doesn't look like it even does anything sensible there.  I don't think it's expected to be used in combination with YSort.  The use-cases are quite different.

Below the YSort node, `behind_parent` nodes otherwise work fine for the sub-children when making a character out of several Sprites.  So only that behavior would need to be preserved.

When comparing against the node to be ysorted, I imagine the best that could be done is to track the final render order/index of the ysort parent, and at some point during the algorithm inserting these `behind_parent` child nodes just above the ysort parent.  IE:  `insert(parent_index - 1, child)`.

The order of the child traversal would be something to watch for as well.

**Comment by andriyDev** [comment]
In either implementation, `behind_parent` works for non-Y-sorted nodes (including ones that are part of some Y-sort tree above its parent). I think trying to force `behind_parent` to do something for Y-sorted nodes isn't sensible. It'll just make the behaviour unclear and I for the life of me cannot think of a situation where you'd need to Y-sort but also remain beneath the parent? The workaround would be to make a root node above the parent and add a new Ysorted tree adjacent to the parent. Trying to jam `behind_parent` behaviour in would almost certainly make the code more complex for something probably no one is going to use. That's my opinion at least, I'd like to see what the others think on this. :)

**Comment by reduz** [comment]
I am starting to see more than one case for 4.0 where not only we need to replace one node type by another, but also do some custom configuration or logic to keep compatibility, so this may require changes to the format loaders. Let me discuss this a bit with other contribs, so maybe we can add a compat API.

**Comment by golddotasksquestions** [comment]
Hi, I would like to leave a comment here from the my perspective as someone who cares about usability first and features only second to that.

2D sorting is already not easy to comprehend for newcomers. It took me ages to make sense of it and it's still difficult for me to explain all the different options and things that may or may not cross influence each other and eventually have a final say in what is drawn on top of what other thing. People come to the community forums on a daily basis looking for help with sorting issues.

I recently [posted a comment on Godot's subreddit](https://www.reddit.com/r/godot/comments/mq9tws/how_do_i_layer_my_nodes/guhockr?utm_source=share&utm_medium=web2x&context=3), trying to explain 2D sorting in Godot, and it took a whole a page long wall of text. Thanks to the changes in 3.1, the current y-sorting solution is one of the easiest things to explain and understand out of all the sorting options.

I think what I am asking for here is to consider if adding this new y-sorting really **makes sorting in Godot easier to understand and more intuitive to use**. 
If it does not fulfill this criteria, but instead adds functionality for the cost of usability, I do think it would be fare to discuss if it is really worth it, and what could be done to make it at least as easy to use and explain as our current feature.

In the vast majority of my time using Godot, I use Ysort extensively, because I love top down 2D games. In 2.5 years of using Ysort in Godot, even in bigger projects, I have **never once** had the need to Ysort the descendants of Ysort children. Nesting Ysort nodes covered all my needs ([in cases like to tidy up the Scene Tree, to group nodes together](https://github.com/godotengine/godot-docs/issues/2286))

I would also welcome a feature change a lot that adds more flexibility and functionality as this one here. Maybe this opens up a possibility for usecases I can't even imagine right now. However I'm also sensing it might result in the same confusion as the current pause implementation periodically brings up, or the Control Container nodes influencing all it's children, or mouse filtering of Control nodes. 

If we start to give nodes like Sprites and KinematicBody2Ds a ysort property that if enabled will ysort all their decendance, this just opens a can of worms. What is I only want some of the children ysorted? Would I have to manually tick every one of them? What if I want all but a few of the children ysorted? Y-sorting the children of a character is almost never wanted for example. Would I need to uncheck the Character Kinematic body then? If I did would that mean I would need a parent node that has this property checked (this is the behavior we currently have btw)? What about the Sprite chain as @andriyDev described? In this case you would want the root to y-sort itself.

The way I see it this will quickly descend into into a similar system as the pause propagation and the mouse filtering. Do we really want that? Who really has a uscase/need for this new y-sorting feature anyway? I have not come across any proposal or community post asking for it or expressing a need. Also people here are cheering, but noone actually explained why this is needed in the first place. It certainly does not close https://github.com/godotengine/godot-docs/issues/2286, instead it raises more questions. It seems like an awesome feature addition on first glance, but I'm afraid it might make things a lot more complicated for very little benefit and without much (if any) need. I would really love to hear about a practical usecase (Isn't this what the [GIP system](https://github.com/godotengine/godot-proposals) is for?)

**Comment by Pennycook** [comment]
I'm inclined to agree with @golddotasksquestions.  It took me a while (and some experiments) to figure out how the current implementation of YSort works, and this new change seems like it would make things even more complicated for new users.

Reading back over the discussion, I'm also beginning to wonder whether there's a good reason to make this a breaking change.  If many people are happy with the existing behavior but other people are interested in some new use-cases, why not make the existing YSort node more configurable instead of changing the default behavior to the most complex and expensive use-case?

For example, the "Enabled" checkbox in the YSort node itself could be augmented by a drop-down with options like "Children", "Descendants" and "Group".  That would allow the existing behavior to remain the default while also providing users a way to opt-out of deep recursive sorts or opt-in to sorting nodes at arbitrary depth (i.e. all nodes in a user-specified group).

File: scene/2d/node_2d.cpp
Comment (akien-mga) [inline scene/2d/node_2d.cpp:490]: We were discussed with some contributors that it would be good to group all sorting related properties together, so instead of the current "Z Index" and "YSort", you'd have a single "Sorting" group for all three properties.

**Comment by akien-mga** [comment]
We discussed this in a PR review meeting yesterday and confirmed that this is a good change. Arguably, making Y sorting a property of `Node2D` instead of a dedicated node might make things more user-friendly, as the property will likely be easier to find than a dedicated node (unless you're experienced and know to search for YSort to do Y sorting).

Could you rebase the PR? Note that the TileMap code was fully rewritten so you might need to adjust the code and make sure that it still works as intended.

@golddotasksquestions Regarding the first part of your comment, there is no change in functionality in this PR. The only thing it changes is that instead of using a `YSort` node to do Y sorting, you'll use a `Node2D` node (or any node that inherits Node2D) and enable the `y_sort_enabled` property.

For further features as discussed by @golddotasksquestions and @Pennycook, you should write a proposal on https://github.com/godotengine/godot-proposals. More options for Y sorting can be considered but that's orthogonal to this PR which is just a refactoring and simplification of the API.

**Comment by Pennycook** [comment]
> For further features as discussed by @golddotasksquestions and @Pennycook, you should write a proposal on https://github.com/godotengine/godot-proposals. More options for Y sorting can be considered but that's orthogonal to this PR which is just a refactoring and simplification of the API.

Sorry about that -- I didn't mean to make an official proposal, just to point out that it seemed there were alternative implementations that wouldn't require the Y-Sort node to be removed.

I think I might be confused about what this change does, because reading back over the discussion makes it seem like much more than a "refactoring and simplification of the API".  There are two things I don't understand:

1) Whether or not a node with Y-Sort enabled is itself included in the Y-Sort.

2) When the "deep Y-Sort" stops.

File: scene/2d/node_2d.cpp
Comment (andriyDev) [reply scene/2d/node_2d.cpp]: Done.

**Comment by andriyDev** [comment]
@Pennycook No worries!

1. A node with Y-Sort enabled is indeed included in the Y-Sort. Practically I don't expect this to come up much, the root of a y-sorting group is likely not to be a renderable element anyway. Regardless, it will work in that situation.
2. Yes, the new y-sort will stop collecting children the first time it encounters a Node2D without y-sorting - just like the existing behaviour!

The codebase can be tricky haha, I came back to this PR and I was in shambles trying to rebase this.

**Comment by andriyDev** [comment]
@akien-mga Done. Rebased, and verified everything is working as intended!

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by Pennycook** [comment]
> @Pennycook No worries!
> 
> 1. A node with Y-Sort enabled is indeed included in the Y-Sort. Practically I don't expect this to come up much, the root of a y-sorting group is likely not to be a renderable element anyway. Regardless, it will work in that situation.
> 2. Yes, the new y-sort will stop collecting children the first time it encounters a Node2D without y-sorting - just like the existing behaviour!
> 
> The codebase can be tricky haha, I came back to this PR and I was in shambles trying to rebase this.

Thanks for the explanation, and sorry again for the noise!

**Pull request merged**
